core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -1197,6 +1213,78 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     KVUtils.open(newStorePath, metadata)
   }
 
+  private def createHybridStore(
+      dm: HistoryServerDiskManager,
+      appId: String,
+      attempt: AttemptInfoWrapper,
+      metadata: AppStatusStoreMetadata): KVStore = {
+
+    var retried = false
+    var hybridStore: HybridStore = null
+    while (hybridStore == null) {
+      val reader = EventLogFileReader(fs, new Path(logDir, attempt.logPath),
+        attempt.lastIndex)
+      val isCompressed = reader.compressionCodec.isDefined
+
+      // Throws an exception if the memory space is not enough
+      memoryManager.lease(appId, attempt.info.attemptId, reader.totalSize, isCompressed)
+
+      logInfo(s"Leasing disk manager space for app $appId / ${attempt.info.attemptId}...")
+      val lease = dm.lease(reader.totalSize, isCompressed)
+      val isLeaseRolledBack = new java.util.concurrent.atomic.AtomicBoolean(false)
+      var store: HybridStore = null
+      try {
+        store = new HybridStore()
+        val levelDB = KVUtils.open(lease.tmpPath, metadata)
+        store.setLevelDB(levelDB)
+        rebuildAppStore(store, reader, attempt.info.lastUpdated.getTime())
+
+        // Start the background thread to dump data to levelDB when writing to
+        // InMemoryStore is completed.
+        store.switchToLevelDB(new HybridStore.SwitchToLevelDBListener {
```

Comment:
I guess this can be moved out of try statement - so that we can treat the error cases during replay and switch separately. Maybe this can also address your comment in below.

Reply:
I would prefer to keep store.switchToLevelDB() in the try block and remove isLeaseRolledBack variable.  If we move it out of the try block, we would need one more variable to track whether we have encountered exceptions during the rebuildAppStore(), otherwise, store.switchToLevelDB() will be executed when the rebuildAppStore() is failed due to IOException. What do you think?

I added a closed variable to indicate whether the hybrid store is closed, and not to start the background thread when the closed is true. This can address the issue above.

Reply:
The reason to split is that retrial logic is only valid for replaying - that said, you don't need to prepare Level DB and deal with disk lease (acquire/rollback) while replaying. You just need to deal with it "after" replaying which should simplify the logic.

Reply:
yeah, that makes sense, I will update the code.